File: Linear_Queue.py
class Queue:
    def __init__(self, max_size):
        self.queue = []
        self.max_size = max_size

    def is_full(self):
        return len(self.queue) == self.max_size

    def is_empty(self):
        return len(self.queue) == 0

    def enqueue(self, element):
        if self.is_full():
            print("Queue is full! Cannot enqueue.")
        else:
            self.queue.append(element)
            print(f"Enqueued: {element}")

    def dequeue(self):
        if self.is_empty():
            print("Queue is empty! Cannot dequeue.")
        else:
            element = self.queue.pop(0)
            print(f"Dequeued: {element}")
            return element

    def display(self):
        if self.is_empty():
            print("Queue is empty!")
        else:
            print("Queue elements:", " -> ".join(map(str, self.queue)))


def main():
    max_size = int(input("Enter the maximum size of the queue: "))
    queue = Queue(max_size)

    while True:
        print("\nQueue Menu:")
        print("1. Enqueue")
        print("2. Dequeue")
        print("3. Display")
        print("4. Exit")
        choice = input("Enter your choice: ")

        if choice == "1":
            element = input("Enter the element to enqueue: ")
            queue.enqueue(element)
        elif choice == "2":
            queue.dequeue()
        elif choice == "3":
            queue.display()
        elif choice == "4":
            print("Exiting program.")
            break
        else:
            print("Invalid choice. Please try again.")


if __name__ == "__main__":
    main()


# # Explanation:-

# Queue Class:

# Implements a queue using a Python list.

# Contains methods for:
# is_full(): Checks if the queue has reached its maximum size.
# is_empty(): Checks if the queue is empty.
# enqueue(element): Adds an element to the rear of the queue if not full.
# dequeue(): Removes and returns the element from the front of the queue if not empty.
# display(): Prints all elements in the queue.

# Menu-Driven Approach:

# Prompts the user with options to perform queue operations.
# Handles invalid choices and exits cleanly when the user selects "Exit".

# Dynamic Input:

# The queue size is set dynamically based on user input at the start of the program.

File: Linear_Queue.c
//Queue implementation using array

#include<stdio.h>
#include<conio.h>
#include<stdlib.h>
#define MAX 3

void insert(); //To insert the data from the linear queue
void delete(); //To delete the data from the linear queue
void display(); //To display the data from the linear queue

int queue[MAX]; //To create queue named queue
int rear = 0;	//To point the back of the queue
int front = 0;  //To point the front of the queue

int main()
{
	int choice;
	while(1) //To display the menu driven program
	{
		printf("\n\n1.Insert element to queue (enqueue)");
		printf("\n2.Delete element from queue(dequeue)");
		printf("\n3.Display all elements of queue");
		printf("\n4.Exit");
		printf("\n");
		printf("\nEnter your choice : ");
		scanf("%d", &choice);
		switch(choice)
		{
			case 1:
				insert();
				break;
			case 2:
				delete();
				break;
			case 3:
				display();
				break;
			case 4:
                printf("\nProgram Exited Successfully...\n\n");
				exit(1);
			default:
				printf("\nInvalid choice");
				break;
		}
	}
	getch();
	return 0;
}

void insert()
{
	int item;
	if(rear == MAX)
		printf("\n\n\tQueue is full, can't add more elements\n\n");
	else
	{
		printf("\nInsert the element in queue : ");
		scanf("%d", &item);
		queue[rear] = item;
		printf("\n%d is inserted in the queue\n",item);
		rear++;	
	}
}

void delete()
{
	if(front == rear)
	{
		printf("\n\n\tQueue Underflow, no elements to delete\n\n");
		return;
	}
	else
	{
		printf("\n%d is deleted from queue....", queue[front]);
		front = front + 1;
	}
}

void display() //To display all the elements of the queue
{
	int i;
	if(front == rear)
		printf("\n\n\tQueue is empty, nothing to display\n\n");
	else
	{
		printf("Elements in Queue : ");
		for(i = front; i < rear; i++)
			printf("%d ", queue[i]);
		printf("\n");
	}
}	
